assets/js/shared/context/product-data-query.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ */
+import { addQueryArgs } from '@wordpress/url';
+import apiFetch from '@wordpress/api-fetch';
+import { useEffect, useState, useRef } from '@wordpress/element';
+import { difference, concat } from 'lodash';
+
+/**
+ * Rendered by Product Data Context to query data for registered fields once children are rendered.
+ */
+const ProductDataQuery = ( { productId, fields, setProduct } ) => {
+	const [ productData, setProductData ] = useState( {} );
+	const retrievedFields = useRef( [] );
+
+	// If the product ID changes, reset fields.
+	useEffect( () => {
+		retrievedFields.current = [];
+		setProductData( {} );
+	}, [ productId ] );
+
+	// When product data changes, update the product context.
+	useEffect( () => {
+		setProduct( productData );
+	}, [ productData, setProduct ] );
+
+	// If the fields change, fetch new data.
+	useEffect( () => {
+		if ( ! productId || fields === undefined ) {
+			return;
+		}
+		const fieldsToFetch = difference( fields, retrievedFields.current );
+
+		if ( fieldsToFetch.length === 0 ) {
+			return;
+		}
+
+		apiFetch( {
+			path: addQueryArgs( `/wc/store/products/${ productId }`, {
+				_fields: [ ...fieldsToFetch, 'id' ],
+			} ),
+		} ).then( ( theProduct ) => {
+			setProductData( ( prevState ) => {
+				return { ...prevState, ...theProduct };
```

Comment:
So I'm wondering if there's potential danger here if in between requests old fields in the state might have changed on the server, but if they weren't requested in the latest request then there's a mix of old and new data?

Reply:
There is some degree of risk if requests are spread apart, but I believe in our use case these requests are likely to be close. e.g. in testing I had the add to cart form do a second product request once the first request had determined product type.

Given the API responses are cached this could already be an issue right?